src/nvme/linux.c

```diff
@@ -474,6 +475,143 @@ int nvme_gen_dhchap_key(char *hostnqn, enum nvme_hmac_alg hmac,
 	memcpy(key, secret, key_len);
 	return 0;
 }
+
+static unsigned char *derive_nvme_keys(const char *hostnqn,
+				       const char *identity, int hmac,
+				       unsigned char *key, int key_len)
+{
+	nvme_msg(NULL, LOG_ERR, "TLS key operations not supported; "\
+		 "recompile with OpenSSL support.\n");
+	return NULL;
+}
+#else
+static unsigned char *derive_retained_key(const EVP_MD *md, const char *hostnqn,
+					  unsigned char *generated_key,
+					  size_t key_len)
+{
+	unsigned char *retained_key;
+	EVP_PKEY_CTX *ctx;
+	size_t retained_len;
+	int err = -ENOKEY;
+
+	retained_key = malloc(key_len);
+	if (!retained_key)
+		return NULL;
+
+	retained_len = key_len;
+	ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
+	if (!ctx)
+		goto out_free_retained_key;
+
+	if (EVP_PKEY_derive_init(ctx) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_set_hkdf_md(ctx, md) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_set1_hkdf_key(ctx, generated_key, key_len) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx,
+					(const unsigned char *)"tls13 ", 6) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx,
+					(const unsigned char *)"HostNQN", 7) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx,
+					(const unsigned char *)hostnqn, strlen(hostnqn)) <= 0)
+		goto out_free_retained_key;
+
+	if (EVP_PKEY_derive(ctx, retained_key, &retained_len) > 0)
+		err = 0;
+
+out_free_retained_key:
+	if (err) {
+		nvme_msg(NULL, LOG_DEBUG,
```

Comment:
Couldn't just make the function to return the error code and the key is part of the argument list?

Because printing to stderr in the library should be avoided unless very very important.

I understand these errors here could happen during normal operation, so we really should return proper error codes which can be presented to the user.

Reply:
Indeed error handling is ugly in C and it leads to compromises in the API design.

Reply:
Reworked functions to return 0/-1 and setting errno.